Load config history (#472)
ServiceNow/azimuth: docs/docs/development/launching.md

* You can consult the openapi documentation at `localhost:8091/docs`. From there, you can consult the API documentation and try out the different endpoints. This can be useful for debugging.
* Note that the back end will not reload automatically based on code changes.

After a successful start, Azimuth saves the provided config in its `config_history.jsonl` artifact. If you use the API to edit the config, the edits are saved there. If you restart Azimuth (for example after shutting it down for the night), you can resume where you left off with:
```shell
make LOAD_CONFIG_HISTORY=1 launch.local
```
In fact, it is possible to specify both `LOAD_CONFIG_HISTORY=1` and a `CFG_PATH` together, in which case Azimuth will automatically

1. load the config from `CFG_PATH` when it first starts (if `config_history.jsonl` is empty); and
2. load the config from `config_history.jsonl` from then on if it is restarted.

Although confusing, this comes in very handy in some contexts like a docker container that might be stopped and restarted with the same command. For example:
```shell
make CFG_PATH=local_configs/development/clinc_dummy/conf.json \
     LOAD_CONFIG_HISTORY=1 \

>      launch.local

Reviewer: In a Docker container, we wouldn't use `launch.local` right?

Author: Correct. In our docker containers, we don't use `make`.
In my commit to [Document LOAD_CONFIG_HISTORY in Getting Started section](https://github.com/ServiceNow/azimuth/pull/472/commits/eba396781a41637805cda9431f987513ecc55c5f), where we use docker, I give those examples with `launch` instead of `launch.local`.

Reviewer: So perhaps we should remove it here too since this looks like an example for a Docker container? At least this was my interpretation.

Author: This is an example that might be useful during local "development". While I am developing this feature, I am running exactly this command. I thought it might be useful to others in the future, but it might be rarer.
The following warning admonition refers to this example
> specify both `LOAD_CONFIG_HISTORY=1` and a `CFG_PATH` together, like in the example above"

so do you think I should move this example inside the admonition?

Reviewer: Are we talking about the same one? I'm talking about the one that follows "like a docker container that might be stopped and restarted with the same command. For example:"

Author: Yes

Author: I think I get what is confusing now. This example is not only about docker, it is about this whole section saying that "it is possible to specify both `LOAD_CONFIG_HISTORY=1` and a `CFG_PATH` together". I am just mentioning that "this comes in very handy in some contexts like a docker container".
So, I think I should move this mention about docker after the example, to avoid this confusion. Does that make sense?
``````diff
diff --git a/docs/docs/development/launching.md b/docs/docs/development/launching.md
index 8ecee59f..06b7eb0f 100644
--- a/docs/docs/development/launching.md
+++ b/docs/docs/development/launching.md
@@ -74,12 +74,13 @@ In fact, it is possible to specify both `LOAD_CONFIG_HISTORY=1` and a `CFG_PATH`
 1. load the config from `CFG_PATH` when it first starts (if `config_history.jsonl` is empty); and
 2. load the config from `config_history.jsonl` from then on (if Azimuth is restarted).

-Although confusing, this comes in very handy in some contexts like a docker container that might be stopped and restarted with the same command. For example:
+For example:
 ```shell
 make CFG_PATH=local_configs/development/clinc_dummy/conf.json \
      LOAD_CONFIG_HISTORY=1 \
      launch.local
 ```
+Although confusing, this comes in very handy in some contexts like a docker container that might be stopped and restarted with the same command.

 ??? warning "Using `LOAD_CONFIG_HISTORY=1` with a custom `artifact_path`"
     The location of `config_history.jsonl` depends on `artifact_path`, so if you customized it in your config, it needs to be consistent when you restart Azimuth. The easiest solution is to always specify both `LOAD_CONFIG_HISTORY=1` and a `CFG_PATH` together, like in the example above. Alternatively, you can specify the environment variable `ARTIFACT_PATH`:
diff --git a/docs/docs/getting-started/c-run.md b/docs/docs/getting-started/c-run.md
index 40eeaebe..d097e156 100644
--- a/docs/docs/getting-started/c-run.md
+++ b/docs/docs/getting-started/c-run.md
@@ -117,10 +117,11 @@ In fact, it is possible to specify both `LOAD_CONFIG_HISTORY=1` and a `CFG_PATH`
 1. load the config from `CFG_PATH` when it first starts (if `config_history.jsonl` is empty); and
 2. load the config from `config_history.jsonl` from then on (if Azimuth is restarted).

-Although confusing, this enables you to stop and restart the docker container with the same command. For example:
+For example:
 ```shell
 make CFG_PATH=/config/my_project/conf.json LOAD_CONFIG_HISTORY=1 launch
 ```
+Although confusing, this enables you to stop and restart the docker container with the same command.
``````

Reviewer: yes, that works!! Thank you!

* You can consult the openapi documentation at `localhost:8091/docs`. From there, you can consult the API documentation and try out the different endpoints. This can be useful for debugging.
* Note that the back end will not reload automatically based on code changes.

> After a successful start, Azimuth saves the provided config in its `config_history.jsonl` artifact. If you use the API to edit the config, the edits are saved there. If you restart Azimuth (for example after shutting it down for the night), you can resume where you left off with `LOAD_CONFIG_HISTORY=1`. In order to find the right `config_history.jsonl`, Azimuth needs the same `ARTIFACT_PATH`, so to resume the above example, it needs to be specified:

Reviewer: We will need to add these details to the user section too: https://servicenow.github.io/azimuth/main/getting-started/c-run/

Author: Thank you. In this commit, I [Document LOAD_CONFIG_HISTORY in Getting Started section](https://github.com/ServiceNow/azimuth/pull/472/commits/eba396781a41637805cda9431f987513ecc55c5f).

```shell

> make ARTIFACT_PATH=cache LOAD_CONFIG_HISTORY=1 launch.local

Author: In our Python code, we could consider making `"cache"` the default artifact path, useful locally, rather than `"/cache"`, useful in docker, and set it to `"/cache"` in our `Dockerfile`.
```suggestion
make LOAD_CONFIG_HISTORY=1 launch.local
```

Reviewer: Yeah, looks like a good idea

Author: I like it. It also makes the documentation more straightforward; a simpler description of the "normal" use case with the default `artifact_path`, and then a warning box about customizing `artifact_path`.

<img width="880" alt="image" src="https://user-images.githubusercontent.com/8386369/222824245-b9f2c991-5327-4cb0-a343-1e326d08c543.png">

```

> In fact, it is possible to specify both `LOAD_CONFIG_HISTORY=1` and a `CFG_PATH` together, in which case Azimuth will automatically

Reviewer: ```suggestion
In fact, it is possible to specify both `LOAD_CONFIG_HISTORY=1` and a `CFG_PATH` together, in which case Azimuth will automatically derive the `artifact_path` from the `CFG_PATH` to find the config history, if available. Azimuth will then:
```
Is that accurate?

Author: Yes! The user might still set an env var `ARTIFACT_PATH`, and not in `CFG_PATH`, but that might be an insignificant detail. Here is a try anyway:
```suggestion
In fact, it is possible to specify both `LOAD_CONFIG_HISTORY=1` and a `CFG_PATH` together. Azimuth will derive the `artifact_path` from `CFG_PATH` and `ARTIFACT_PATH` to find the right `config_history.jsonl`. Then, Azimuth will automatically:
```

Reviewer: Looks good!

1. load the config from `CFG_PATH` when it first starts (if `config_history.jsonl` is empty); and

> 2. load the config from `config_history.jsonl` from then on if it is restarted.

Author: I think I'll clarify that "it" is "Azimuth" here.
```suggestion
2. load the config from `config_history.jsonl` from then on (if Azimuth is restarted).
```

Although confusing, this comes in very handy for a docker container that might be stopped and restarted with the same command. For example:
```shell
make CFG_PATH=local_configs/development/clinc_dummy/conf.json \
     LOAD_CONFIG_HISTORY=1 \
     launch.local

> ```

Reviewer: Very clear, thank you!
